File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/FixedSizedRecords.pl
#!/usr/bin/perl -w
use strict;

### FixedSizedRecords.pl

use constant RECORD_LEN => 32;                                       #(A)

# Create a small database of fixed-sized records:
open(FILE, ">data2.db");    # die stuff                              #(B)
print FILE pack( "a10 a10 a12","kak","avi","765-123-4567" );         #(C)
print FILE pack( "a10 a10 a12","smythe","stacey","765-234-6666" );   #(D)
print FILE pack( "a10 a10 a12","beeblecrox","zaphod","111-987-9988");#(E)
print FILE pack( "a10 a10 a12","skywalker","luke","4ZV 398 9999" );  #(F)
close FILE;                                                          #(G)

# Open the database for user interaction:

open(FILE, "+< data2.db") or die "unable to open file: $!";          #(H)
my ($buff, %names);                                                  #(I)

# Construct the list of all the last names:
for (my $i=0;;$i++) {                                                #(J)
    my $file_pointer = $i * RECORD_LEN;                              #(K)
    seek(FILE, $file_pointer, 0);                                    #(L)
    last unless read(FILE, $buff, 10);                               #(M)
    $buff =~ s/\0*$//;                                               #(N)
    $names{$buff} = $i;                                              #(O)
}

# User interaction:
while (1) {                                                          #(P)
    print "Last names of the individuals in the database:\n\n";      #(Q)
    print map "$_\n", sort keys %names;                              #(R)
    print "\nWhose record would you like to see? Enter last name: "; #(S)
    chomp( my $response = <STDIN> );                                 #(T)
    last unless exists $names{$response};                            #(U)
    my $row_index = $names{$response};                               #(V)
    seek(FILE, $row_index * RECORD_LEN, 0);                          #(W)
    read(FILE, $buff, 32);                                           #(X)
    my ($lname, $fname, $phone) = unpack( "a10 a10 a12", $buff );    #(Y)
    printf "\nHere it is:    %s\t%s\t%s\n", $lname, $fname, $phone;  #(Z)
    printf "\nEnter the old value you'd like to change: ";           #(a)
    chomp( my $old = <STDIN> );                                      #(b)
    printf "\nNow enter the new value: ";                            #(c)
    chomp( my $new = <STDIN> );                                      #(d)
    foreach ($lname, $fname, $phone) {                               #(e)
        last if s/$old/$new/;                                        #(f)
    }
    seek(FILE, $row_index * RECORD_LEN, 0);                          #(g)
    print FILE pack( "a10 a10 a12", $lname, $fname, $phone );        #(h)
}

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/FixedSizedRecords.py
#!/usr/bin/env python

###  FixedSizedRecords.py

import struct, sys                                                   #(A)

RECORD_LEN = 32                                                      #(B)
FILE = None                                                          #(C)

# Create a small database of fixed-sized records:
try:                                                    
    FILE = open("data2.db", "w")                                     #(D)
    FILE.write( struct.pack("10s 10s 12s","kak",\
                                              "avi","765-123-4567")) #(E)
    FILE.write(struct.pack("10s 10s 12s","smythe",\
                                           "stacey","765-234-6666")) #(F)
    FILE.write(struct.pack("10s 10s 12s","beeblecrox",\
                                           "zaphod","111-987-9988")) #(G)
    FILE.write(struct.pack("10s 10s 12s","skywalker",\
                                             "luke","4ZV 398 9999")) #(H)
except IOError, inst:                           
    print "Error in file IO: ", inst                                
if FILE: FILE.close()                                                #(I)

# Open the database for user interaction:
names = {}                                                           #(J)
try:                                                                 #(K)
    # Open the file in read/write mode:
    FILE = open("data2.db", "r+")                                    #(L)
    i = 0
    while 1:                                                         #(M)
        file_pointer = i * RECORD_LEN                                #(N)
        FILE.seek( file_pointer, 0 )                                 #(O)
        last_name = FILE.read( 10 )                                  #(P)
        if not last_name: break                                      #(Q)
        last_name = last_name.strip("\x00")                          #(R)
        names[last_name] = i                                         #(S)
        i += 1
except IOError, inst:        
    print "Error in creating a name-row_index dictionary: ", inst      

while 1:                                                             #(T)
    print "\nLast names of the individuals in the database:\n"       #(U)
    for item in sorted( names.keys() ): print item                   #(V)
    response = raw_input(\
        "\nWhose record would you like to see? Enter last name: " )  #(W)
    if not names.has_key( response ): break                          #(X)
    row_index = names[ response ]                                    #(Y)
    FILE.seek( row_index * RECORD_LEN, 0 )                           #(Z)
    buff = ''
    try:
        buff = FILE.read( 32 )                                       #(a)
    except IOError, inst:
        print "read error: ", inst
        sys.exit(1)
    record = list( struct.unpack( "10s 10s 12s", buff ) )            #(b)
    print "\nHere it is:    %s\t%s\t%s" % \
                            (record[0], record[1], record[2])        #(c)
    old = raw_input("\nEnter the old value you'd like to change: ")  #(d)
    new = raw_input( "\nNow enter the new value: " )                 #(e)
    for i in range( len(record) ):                                   #(f)
        if record[i].strip("\x00") == old: record[i] = new; break    #(g)
    if i == 0:                                                       #(h)
        print( "illegal change: changing last name not allowed" )    #(i)
        continue                                                     #(j)
    try:
        FILE.seek( row_index * RECORD_LEN, 0 )                       #(k)
        FILE.write( struct.pack( "10s 10s 12s", \
                           record[0], record[1], record[2] ) )       #(l)
    except IOError, inst:
        print "write error: ", inst
        sys.exit(1)

if FILE: FILE.close()                                                #(m)
